PR: Move Pylint plugin to new API

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))


<!--- Explain what you've done and why --->

Took the opportunity to simplify the buttons so that it is similar to find in files (start/stop are the same button)

![pylint](https://user-images.githubusercontent.com/3627835/78578410-1e768c80-77f5-11ea-92e1-f1a65e593194.gif)



### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Depends on #11741
Fixes #12183


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @goanpeca! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`spyder/api/widgets.py`](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/api/widgets.py):

> [Line 452:80](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/api/widgets.py#L452): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 696:80](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/api/widgets.py#L696): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)



* In the file [`spyder/plugins/ipythonconsole/plugin.py`](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/ipythonconsole/plugin.py):

> [Line 423:21](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/ipythonconsole/plugin.py#L423): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator











* In the file [`spyder/plugins/pylint/plugin.py`](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/plugin.py):

> [Line 72:80](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/plugin.py#L72): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 102:80](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/plugin.py#L102): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 129:80](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/plugin.py#L129): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 201:1](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/plugin.py#L201): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`spyder/plugins/pylint/widgets/pylintgui.py`](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/widgets/pylintgui.py):

> [Line 173:40](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/widgets/pylintgui.py#L173): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 292:71](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/widgets/pylintgui.py#L292): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 296:71](https://github.com/spyder-ide/spyder/blob/dae929ec9d48a9945729949c1fb1afdb37bc7d46/spyder/plugins/pylint/widgets/pylintgui.py#L296): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

juanis2112 commented:
/binder

juanis2112 commented:
Some comments about this PR:
1. The shortcut F8 is not working for running the code analysis 
2. Clicking on an element in the pane (warning, error, message) doesn't go to the line corresponding to this element. However if I close the file and click the error message in the pane it opens the file (but again it doesn't go to the corresponding line) This was a bug that we found and it was fixed by @ccordoba12  in 4.1.4 so not sure why it is not working here. I tested it in 4x and it works fine.
3. The hamburger menu doesn't have the expand/collapse items that it had before. 
4. When selecting a file in the pane with the "Select python file" button, the analysis doesn't run automatically which it did before. It does display the path of the file selected in the combobox but when pressing analyze, it runs the analysis on the file that is selected in the editor and displays the path of this last file in the combobox. 
5. When reducing the width of the pane it is noticeable that there is some sort of "invisible" item between the analyze button and the hamburger menu button which shouldn't be there. In contrast, the combobox should take the full space when the width is less (as it did before in 4.x)
![Screen Shot 2020-07-30 at 1 28 01 AM](https://user-images.githubusercontent.com/18587879/88888442-ecee5a80-d203-11ea-8e71-b380f3ac062d.png)

goanpeca commented:
Hi @juanis2112, thanks for the review:

>1. The shortcut F8 is not working for running the code analysis

I cannot reproduce this. On my Mac pressing `Fn+F8` run the analysis as expected.

>2. Clicking on an element in the pane (warning, error, message) doesn't go to the line corresponding to this element.

Fixed

>3. The hamburger menu doesn't have the expand/collapse items that it had before.

Fixed

>4. When selecting a file in the pane with the "Select python file" button, the analysis doesn't run automatically which it did before

Fixed

>5. When reducing the width of the pane it is noticeable that there is some sort of "invisible" item between the analyze button

Fixed

goanpeca commented:
>more differences between 4.x and this PR:

Fixed.

ccordoba12 commented:
/show binder

ccordoba12 commented:
> The shortcut F8 is not working for running the code analysis

This is still a problem on Linux and Windows. Please check it on a virtual machine.

All the other problems reported by @juanis2112 and I are solved now. Great job @goanpeca!

goanpeca commented:
>This is still a problem on Linux and Windows. Please check it on a virtual machine.

Could not reproduce but introduced a change that might help. Could you check @juanis2112, @ccordoba12 ?

ccordoba12 commented:
> Could not reproduce but introduced a change that might help.

It's working now. Thanks @goanpeca.

@dalthviz will give this one a final manual review and if he finds everything in order, he'll merge it.

goanpeca commented:
>While testing this I found some things but the only related with this PR is that seems like a min width is being set so is not possible to resize the window (it only allows you to resize up to certain point):

This one is related to this PR, I will fix!

Thanks @dalthviz :-)

ccordoba12 commented:
> This error is also present in 4.1.4 so I will create an issue or maybe already one is open @ccordoba12 ?

Please open a new issue about this one. I don't remember seeing this reported yet.

goanpeca commented:
>While testing this I found some things but the only related with this PR is that seems like a min width is being set so is not possible to resize the window (it only allows you to resize up to certain point):

@dalthviz I could not reproduce on Mac, will check on a Win VM and see if it is a windows thing only.

ccordoba12 commented:
@goanpeca, please address this comment from @dalthviz:

> While testing this I found some things but the only related with this PR is that seems like a min width is being set so is not possible to resize the window (it only allows you to resize up to certain point)

so we can merge it.

goanpeca commented:
>@dalthviz I could not reproduce on Mac, will check on a Win VM and see if it is a windows thing only.

I cannot reproduce this on the windows VM, any particular thing that might be triggering it?

dalthviz commented:
@goanpeca after checking a little bit more seems like the strange behavior is related with the way the toolbars are being contracted (sometimes the resize gets stuck, although I don't now why):

![resize](https://user-images.githubusercontent.com/16781833/90820386-e7fa6380-e2f6-11ea-9b43-2a795a7a6df7.gif)

Looking at the code seems like a custom handling is being implemented for the `resizeEvent` to emit a custom signal: https://github.com/spyder-ide/spyder/blob/a24398264e5cb5dbbb72fda6a35555fd08fb40a4/spyder/api/widgets/__init__.py#L540-L545. Maybe that could be causing the resize to get stuck?

goanpeca commented:
Thanks for checking @dalthviz. Just to check is it associated to the speed in which you move?

>Looking at the code seems like a custom handling is being implemented for the resizeEvent to emit a custom signal:

Correct, that change was needed to be able to have the toolbar items use as much area as possible.

Question: If you do it slowly it does not get stuck, but if you do it faster it does?

goanpeca commented:
@dalthviz made some small changes to see if the situation improves.

Does increasing this number helps?

https://github.com/spyder-ide/spyder/pull/12160/files#diff-f34c043d56467c819418192611e89a6eR421

dalthviz commented:
@goanpeca checked again and seems the problem is still there :/

> Question: If you do it slowly it does not get stuck, but if you do it faster it does?

The behavior is the same

> Does increasing this number helps?

I'm not sure what number is that, is an option?

From what I'm seeing seems like the resize signal is applying only for the first toolbar that isn't contracted from right to left. So, to contract the rest, you need to again start the resize action. Maybe this GIF illustrates that (first the working directory toolbar needs two resize events to contract - probably do the combobox - and then the rest of the toolbars need one resize event each):

![resize](https://user-images.githubusercontent.com/16781833/90831973-70363400-e30a-11ea-8d49-c87726679591.gif)

goanpeca commented:
>From what I'm seeing seems like the resize signal is applying only for the first toolbar that isn't contracted from right to left. So, to contract the rest, you need to again start the resize action. Maybe this GIF illustrates that (first the working directory toolbar needs two resize events to contract - probably do the combobox - and then the rest of the toolbars need one resize event each)

Thanks @dalthviz, I understand what you mean. However the changes  I made refer to the toolbar in the Docks (each pane has now the options as a toolbar)

<img width="477" alt="Screen Shot 2020-08-20 at 17 33 48" src="https://user-images.githubusercontent.com/3627835/90832408-53e6c700-e30b-11ea-979e-3eaf467b1d6e.png">

But it should now have any effect on the application toolbars. It seems like it is a separate issue. On master you see the same behavior?

dalthviz commented:
Checking with master (3bf92535c) I don't see that behavior:

![resize2](https://user-images.githubusercontent.com/16781833/90832915-7decb900-e30c-11ea-9e40-33990611b4a1.gif)

goanpeca commented:
@dalthviz I cannot reproduce on any of the systems (binder linux) local mac and Win 10 VM. It is going to be really hard to try to fix something I cannot reproduce.

Could we add an item to https://github.com/spyder-ide/spyder/issues/13527 and move forward. Seems like a very minor issue to block this PR from merging.

ccordoba12 commented:
> Could we add an item to #13527 and move forward

I don't agree with this. Please set up a Zoom call with @dalthviz so they two of you can fix this issue.

goanpeca commented:
@ccordoba12 the issue stems from trying to address this comment:

<img width="825" alt="Screen Shot 2020-08-20 at 18 48 45" src="https://user-images.githubusercontent.com/3627835/90836736-cc528580-e315-11ea-8de2-320fac86eca0.png">

If I revert the "fix", then the problem is gone, but the "invisible" item is back.

That is a problem that affects all plugins that provide a widget inside the Pane (like the combobox in this case) and it deserves a fix that covers all widgets. It is not a problem specific to this Plugin, even if it was a comment that came out of reviewing it.

--- 

@dalthviz please create an event whenever it suits you. Available hours are visible on the company calendar.

>I don't agree with this. Please set up a Zoom call with @dalthviz so the two of you can fix this issue.

Fair enough, but even if you do not agree with this, @dalthviz should be able to have also an opinion and a vote, after all the question was addressed to him.

dalthviz commented:
> @dalthviz please create an event whenever it suits you. My available hours are visible on my company calendar.

Ok :+1: done

dalthviz commented:
@goanpeca please rebase this one to get the changes done in #13600

goanpeca commented:
>@goanpeca please rebase this one to get the changes done in #13600

Done @dalthviz.

dalthviz commented on spyder/api/widgets/__init__.py:
```diff
@@ -766,7 +766,11 @@ def get_options_menu(self):
         Return the main options button of the widget.
+=======
+        Return the main options menu button of the widget.
+>>>>>>> Migrate Pylint plugin to new API
```
@goapeca seems like the rebase had some issues, please fix this

goanpeca replied:
Oops. Fixing

goanpeca replied:
Fixed @dalthviz

dalthviz commented:
/show binder

dalthviz commented on spyder/plugins/pylint/main_widget.py:
```diff
@@ -0,0 +1,899 @@
+        self.datelabel = QLabel(self)
+        self.treewidget = ResultsTree(self)
+
+        # FIXME: Move to plugin?
```
So, should we move the things below this line to the plugin? (or was this comment before the move to the API maybe?)

goanpeca replied:
Removed the FIXME notice.

goanpeca commented:
>@goanpeca left a question and also I think this plugin should use the spinner. Besides that LGTM 👍

Thanks for the feedback @dalthviz, made the requested fix.

dalthviz commented:
@andfoy @steff456 could you give to this PR a manual test on Linux and Mac? Thanks!

steff456 commented:
/show binder

steff456 commented:
Hi @goanpeca and @dalthviz,

With @andfoy we manually review this PR and everything is working as expected.

dalthviz reviewed (commented):
@goanpeca one last comment, please unify the use of `"` to enclose strings, there is a mix (as an example the pylint `main_widget.py` in the setup of the actions) and using just `"` could improve future readability of the code. Thanks!

goanpeca commented:
>@goanpeca one last comment, please unify the use of " to enclose strings, there is a mix (as an example the pylint main_widget.py in the setup of the actions) and using just " could improve future readability of the code. Thanks!

Fixed @dalthviz.

dalthviz commented on spyder/plugins/pylint/main_widget.py:
```diff
@@ -0,0 +1,899 @@
+        change_history_depth_action = self.create_action(
+            PylintWidgetActions.ChangeHistory,
+            text=_("History..."),
+            tip=_("Set history maximum entries"),
+            icon=self.create_icon('history'),
```
@goapeca seems like the mix of `'` and `"` is still there, maybe the rebase didn't pick the changes?

goanpeca replied:
You re right @dalthviz I pushed the rebase but not the commit for fixing the quotes. My bad.

Fixed noew.

dalthviz reviewed (commented):
@goanpeca seems like no change regarding the mix of `'` and `"` was added with the rebase, could you check what happen?

dalthviz commented on spyder/plugins/pylint/main_widget.py:
```diff
@@ -0,0 +1,899 @@
+                "pylint",
+                "--output-format=text",
+                "--msg-template="
+                ""{msg_id}:{symbol}:{line:3d},{column}: {msg}"",
```
Seems like the replace needs some changes here (this is causing the error in the CIs)

dalthviz replied:
Also, some PEP8 errors raised. I know these kind of things are kind of a pain to change so thanks for the effort :+1:

goanpeca replied:
No worries @dalthviz, it is how it goes.

Thanks for all the reviewing. You are all doing a great job, and practice makes perfect :-) !

dalthviz reviewed (commented):
@goanpeca a comment pointing were the CIs are failing

dalthviz reviewed (commented):
@goanpeca checking seems like the selected path in the combobox is not being taken in to account unless you use the `Select python file` button to update it:

![combopylint](https://user-images.githubusercontent.com/16781833/92145509-ef813880-eddd-11ea-9fd3-967d857a6335.gif)

Edit: Seems like the `Analyze` button always triggers the action using the current filename in the Editor

goanpeca commented:
>Edit: Seems like the Analyze button always triggers the action using the current filename in the Editor

Bug!

Will fix.

ccordoba12 commented:
Hey @dalthviz, is this one ready for merge?

dalthviz commented:
@ccordoba12 I would say that some testing could be nice, otherwise the missing thing is a squash

ccordoba12 reviewed (approved):
Thanks @goanpeca for your work on this PR and @dalthviz for your help to finish it.

I also gave this work some manual testing too (per @dalthviz's request) and it's working as expected.
